flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java

```diff
@@ -381,17 +383,20 @@ void onConsumedSubpartition(int subpartitionIndex) {
 			return;
 		}
 
-		int refCnt = pendingReferences.decrementAndGet();
-
-		if (refCnt == 0) {
+		if (isManagedExternally) {
 			partitionManager.onConsumedPartition(this);
```

Comment:
if `isManagedExternally`, we should not call `partitionManager.onConsumedPartition(this)` after one sub partition is consumed.

Reply:
It should still be safe to do so (because the RPM basically ignores the call), but yeah this shows that Ih aven't thought this through properly.

Reply:
Could we call `pin()` in `createSubpartitionView` and decrement the refcount as usual in `onConsumedPartition`?

Reply:
The `pin` would increase the counter as the number of subpartitions, but `createSubpartitionView` only indicates for one of the subpartitions. I need think through this issue again. 
Maybe we could focus on the pin issue in a separate ticket and this PR only introduces the flag but not changed any current behaviors if you like.

Reply:
I think I've found a way to do it; pin() will continue to increment the refCount by the number of subpartitions, but createSubpartitionsView will increment it by 1 (if the partition is externally managed).

I'm not sure whether this is a good approach, and what plans people had on how this should be implemented, so I wouldn't oppose deferring this to a follow-up.

Reply:
TBH I do not like the current way because it seems a bit hacky. 
Because during `ResultPartition#onConsumedSubpartition`, the counter would never be 0 for external case, so the `ResultPartitionManager#onConsumedPartition` would never be called, then the internal checking logic in `ResultPartitionManager#onConsumedPartition` seems reductant.

Reply:
Maybe we could always `pin` as before, but migrate the checking logic from `ResultPartitionManger` to `ResultPartition` as I mentioned before. I mean `ResultPartitionManager` is not aware of the external issue, the tag in `ResultPartitionManager#isReleaseExternallyManagedPartitionsOnConsumption` could be merged with `ResultPartition#isManagedExternally` to generate a final tag in `ResultPartition`.

Then while `ResultPartition#onConsumedSubpartition`, it could check the final tag to decide whether to call `ResultPartitionManager#onConsumedPartition`. And the reference counter only works when the final tag is false.

Reply:
In conclusion:

- do not modify previous `ResultPartitionManger`, only add final tag in `ResultPartition`

- pin logic could keep same as before.

- the ref counter check in `ResultPartition#createSubpartitionView` only works when tag false

- the ref counter check in `ResultPartition#onConsumedSubpartition` only works when tag false